### PR TITLE
dbt-materialize: support enforcing contracts for pseudo-types

### PR DESCRIPTION
**Pull request by morsapaes**
As noted in [#23953](https://github.com/MaterializeInc/materialize/pull/23953#issuecomment-1861133197), the `dbt-materialize` adapter is unable to enforce model contracts for pseudo-type columns. Until we fix #3340, one possible workaround is to override the global `cast()` macro to special-handle these types, and then override `get_empty_schema_sql()` to use that macro instead of hardcoding the cast (this will be fixed soon in dbt (see [`dbt-adapters` #165](https://github.com/dbt-labs/dbt-adapters/pull/165)).

### CODE REVIEW

**Comment by morsapaes** [comment]
Ah. I think the global `cast()` macro might only be available from dbt 1.8 (via `dbt-adapters`).

File: misc/dbt-materialize/dbt/include/materialize/macros/columns.sql [hunk 1]
```diff
@@ -22,3 +22,28 @@
...
+      {%- if col['data_type'] is not defined -%}
+        {%- do col_err.append(col['name']) -%}
+      {#-- If this column's type is just 'numeric' then it is missing precision/scale, raise a warning --#}
+      {%- elif col['data_type'].strip().lower() in ('numeric', 'decimal', 'number') -%}
+        {%- do col_naked_numeric.append(col['name']) -%}
+      {%- endif -%}
+      {% set col_name = adapter.quote(col['name']) if col.get('quote') else col['name'] %}
+      -- NOTE(morsapaes): in a future release of dbt, the default macro will use
+      -- the global cast macro, as modified here, and we can remove this custom
+      -- override. See: https://github.com/dbt-labs/dbt-adapters/pull/165
```
Comment (morsapaes) [inline misc/dbt-materialize/dbt/include/materialize/macros/columns.sql:41]: Oh, this was already released in [`dbt-adapters` v1.1.0](https://github.com/dbt-labs/dbt-adapters/releases/tag/v1.1.0). Still getting used to the new package structure. 🫠

**Comment by benesch** [comment]
Ugh, sorry, I think we were on different pages as to what the ask was! The good news, though, is that you might actually have a better handle on what the customer ask is.

Let me explain. What I was envisioning adding support for was strongly typed lists, where you create a named list type like so:

```
create type my_int_list as list (element type = int);
materialize=> select list['1', '2', '3']::my_int_list;
  list   
---------
 {1,2,3}
(1 row)

```

And then you can use it in a contract like so:

```yml
models:
   - name: test_list_types
     config:
       contract:
         enforced: true
     columns:
       - name: c
         data_type: my_int_list
```

Maybe this already works? I haven't had a free moment to test it. 

It'd also be cool to support strongly typed but anonymous lists (`int list`, in this example):

```yml
models:
   - name: test_list_types
     config:
       contract:
         enforced: true
     columns:
       - name: c
         data_type: "int list"
```

Again, maybe this already works? Not sure!

What you've got here, which is plausibly what the customer's asked for, is about weakly typed list/map/record types. That's definitely got its uses, but is a much weaker assertion than a strongly typed list/map/record. If I were in the customer's shoes, I'd be endeavoring to use strongl ytyped list/map/record types everywhere I could, though I can understand wanting the weak typing as an escape hatch.

That said, I'm worried that the implementation here will break other users of the `cast` macro! I'll drop a line comment on the PR in a second with more detail.

tl;dr if strongly typed list/map types already work with the dbt adapter, I think we should push the customer to use those wherever they can. In parallel let's see if we can an engineer a way to get the psuedotypes working that doesn't break the `cast` macro.

File: misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+{% macro materialize__cast(expression, data_type) -%}
+    {#-- Handle types that don't support cast(NULL as type) --#}
+    {%- if expression.strip().lower() == "null" and data_type.strip().lower() == "map" -%}
+      MAP[NULL=>NULL]
```
Comment (benesch) [inline misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql:19]:
```suggestion
      NULL::map[text => text]
```

File: misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    {%- elif expression.strip().lower() == "null" and data_type.strip().lower() == "list" -%}
+      LIST[NULL]
```
Comment (benesch) [inline misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql:21]:
```suggestion
      NULL::list[text]
```

File: misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+    {%- elif expression.strip().lower() == "null" and data_type.strip().lower() == "record" -%}
+      ROW(NULL)
```
Comment (benesch) [inline misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql:23]:
```suggestion
      (select row() where false)
```

There isn't a better way to create a `NULL` of type `record` that I'm aware of :/.

File: misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+
+{% macro materialize__cast(expression, data_type) -%}
+    {#-- Handle types that don't support cast(NULL as type) --#}
+    {%- if expression.strip().lower() == "null" and data_type.strip().lower() == "map" -%}
```
Comment (benesch) [inline misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql:18]: These expressions don't actually return a `NULL` even though the expression is `NULL`, which could break other users of the `cast` macro that are expecting `cast("null")` to return a real `NULL`, rather than just inspecting its type.

I think it's an easy fix though. Suggestions inline.

File: misc/dbt-materialize/dbt/include/materialize/macros/columns.sql [hunk 6]
```diff
@@ -22,3 +22,28 @@
...
+      {{ cast("null", col['data_type']) }} as {{ col_name }}{{ ", " if not loop.last }}
+    {%- endfor -%}
+    {%- if (col_err | length) > 0 -%}
+      {{ exceptions.column_type_missing(column_names=col_err) }}
+    {%- elif (col_naked_numeric | length) > 0 -%}
+      {{ exceptions.warn("Detected columns with numeric type and unspecified precision/scale, this can lead to unintended rounding: " ~ col_naked_numeric ~ "`") }}
+    {%- endif -%}
+{% endmacro %}
```
Comment (benesch) [inline misc/dbt-materialize/dbt/include/materialize/macros/columns.sql:49]: To confirm my understanding: this file will go away after the 1.8 upgrade merges? Are you planning to wait until the 1.8 upgrade merges before merging this PR?

Comment (morsapaes) [reply misc/dbt-materialize/dbt/include/materialize/macros/columns.sql:49]: Yes, because it depends on the new package structure in dbt v1.8. The upgrade PR is ready to merge, with a couple of follow-up TODOs.

Comment (morsapaes) [reply misc/dbt-materialize/dbt/include/materialize/macros/columns.sql:49]: And the version of the macro above was already released in `dbt-adapters`, yeah. This is now released independently to `dbt-core`, after the refactoring on their end.

File: misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql (hunk 5 above)
Comment (morsapaes) [reply misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql:18]: Thank you!

File: misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql (hunk 3 above)
Comment (morsapaes) [reply misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql:21]: Hum, this notation doesn't seem to work:

```sql
materialize=> SELECT NULL::list[text];
ERROR:  Expected right square bracket, found TEXT
LINE 1: select NULL::list[text];
```

But this does (h/t @nrainer-materialize):

```sql
materialize=> SELECT NULL::text list;
 text list
-----------
 null
```

**Comment by morsapaes** [comment]
> What you've got here, which is plausibly what the customer's asked for, is about weakly typed list/map/record types. That's definitely got its uses, but is a much weaker assertion than a strongly typed list/map/record.

That's what I was solving for, yeah. I expect most users to not strongly type records, despite it being the best practice _in theory_. We should support both scenarios, though! I'll give it another pass.

File: misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql (hunk 3 above)
Comment (benesch) [reply misc/dbt-materialize/dbt/include/materialize/macros/utils/cast.sql]: Oh, d'oh! Sorry, I was tired last night.

**Review by benesch — APPROVED** [review]
LGTM modulo the shuffling around the v1.8 release! Thanks for thinking of this trick. It's pretty nifty.

**Comment by morsapaes** [comment]
Fixed in #27011, so closing. As discussed offline, named `map`, `list`, and `record` types are not enforced (but rather match against base type) ; this will be fixed after #27318 is released.
